gson/src/main/java/com/google/gson/GsonBuilder.java

```diff
@@ -728,16 +731,20 @@ public GsonBuilder registerTypeAdapterFactory(TypeAdapterFactory factory) {
    * @param typeAdapter This object must implement at least one of {@link TypeAdapter},
    *        {@link JsonSerializer} or {@link JsonDeserializer} interfaces.
    * @return a reference to this {@code GsonBuilder} object to fulfill the "Builder" pattern
-   * @throws IllegalArgumentException if the Type adapter being registered is for a JsonElements or any of its subclasses
+   * @throws IllegalArgumentException if the type adapter being registered is for {@link JsonElement} or any of its subclasses
    * @since 1.7
    */
   @CanIgnoreReturnValue
   public GsonBuilder registerTypeHierarchyAdapter(Class<?> baseType, Object typeAdapter) {
     Objects.requireNonNull(baseType);
-    $Gson$Preconditions.checkArgument((typeAdapter instanceof JsonSerializer<?>
+    $Gson$Preconditions.checkArgument(typeAdapter instanceof JsonSerializer<?>
         || typeAdapter instanceof JsonDeserializer<?>
-        || typeAdapter instanceof TypeAdapter<?>)
-        && !JsonElement.class.isAssignableFrom(baseType));
+        || typeAdapter instanceof TypeAdapter<?>);
+
+    if(JsonElement.class.isAssignableFrom(baseType)){
```

Comment:
Same as Éamonn's other comment, there should probably be spaces here to match the style guide:
```suggestion
    if (JsonElement.class.isAssignableFrom(baseType)) {
```